src/middlewared/middlewared/plugins/idmap.py

```diff
@@ -23,281 +24,342 @@ class dstype(enum.Enum):
-            data['domain'] = data['domain']['idmap_domain_name']
-        else:
-            domain_info = await self.middleware.call(
-                'idmap.domain.query', [('domain', '=', data['domain']['idmap_domain_dns_domain_name'].upper())]
-            )
-            data['domain'] = domain_info[0]['name']
+    async def idmap_compress(self, data):
+        opt_enums = ['ssl', 'linked_service']
+        if data.get('options'):
+            for i in opt_enums:
+                if data['options'].get(i):
+                    data['options'][i] = data['options'][i].lower()
+
+        data['idmap_backend'] = data['idmap_backend'].lower()
 
         return data
 
     @private
-    async def _validate_domain_info(self, domain, verrors):
+    async def get_next_idmap_range(self):
         """
-        Only domains that have been configured as idmap domains
-        are permitted.
+        Increment next high range by 100,000,000 ids. This number has
+        to accomodate the highest available rid value for a domain.
+        Configured idmap ranges _must_ not overlap.
         """
-        configured_domains = await self.middleware.call('idmap.domain.query')
-
-        id = domain.get('id')
-        id_verified = False
+        domains = await self.query()
+        sorted_idmaps = sorted(domains, key=lambda domain: domain['range_high'])
+        low_range = sorted_idmaps[-1]['range_high'] + 1
+        high_range = sorted_idmaps[-1]['range_high'] + 100000000
+        return (low_range, high_range)
 
-        short_name = domain.get('idmap_domain_name')
-        short_name_verified = False
+    @private
+    async def remove_winbind_idmap_tdb(self):
+        sysdataset = (await self.middleware.call('systemdatast.config'))['basename']
+        ts = str(datetime.datetime.now(datetime.timezone.utc).timestamp())[:10]
+        await self.middleware.call('zfs.snapshot.create', {'dataset': f'{sysdataset}/samba4',
+                                                           'name': f'wb_cache-{ts}'})
+        try:
+            os.remove('/var/db/system/samba4/winbindd_idmap.tdb')
+        except Exception:
+            self.logger.debug("Failed to remove winbindd_idmap.tdb.", exc_info=True)
 
-        for d in configured_domains:
-            if d['id'] == id:
-                id_verified = True
-            if d['name'] == short_name:
-                short_name_verified = True
+    @accepts()
+    @job(lock='clear_idmap_cache')
+    async def clear_idmap_cache(self, job):
+        """
+        Stop samba, remove the winbindd_cache.tdb file, start samba, flush samba's cache.
+        This should be performed after finalizing idmap changes.
+        """
+        await self.middleware.call('service.stop', 'cifs')
 
-        if id is not None and not id_verified:
-            verrors.add('domain.id', f'Domain [{id}] does not exist.')
+        try:
+            os.remove('/var/db/system/samba4/winbindd_cache.tdb')
```

Comment:
Same question here

Reply:
This is potentially useful information. Will log simple message (non-traceback) if FileNotFound.  Get traceback if it's something really unexpected.